[SPARK-5672][Web UI] Don't return `ERROR 500` when have missing args

catap opened the pull request:
Spark web UI return `HTTP ERROR 500` when GET arguments is missing.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
I agree that 5xx means server error, and is not an appropriate response. But assuming a value of -1 doesn't seem like a great solution -- how about returning `400 Bad Request`?

catap commented:
Ok, the patch was updated.

srowen commented on core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala:
```diff
@@ -32,7 +32,12 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
   def render(request: HttpServletRequest): Seq[Node] = {
     listener.synchronized {
-      val jobId = request.getParameter("id").toInt
+      val paramaterId = request.getParameter("id")
```
Nit: paramater -> parameter

srowen commented on core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala:
```diff
@@ -32,7 +32,12 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
     listener.synchronized {
-      val jobId = request.getParameter("id").toInt
+      val paramaterId = request.getParameter("id")
+      if (paramaterId.equals("")) {
```
It is `null` if not specified, right?

catap replied:
Nope, is return "" when not specified.

srowen replied:
@catap Hm, no it looks like it is in fact `null`: http://docs.oracle.com/javaee/6/api/javax/servlet/ServletRequest.html#getParameter(java.lang.String)

catap replied:
@srowen, yes, you are right.

But for case `&id=` it's return "". And I used this case for testing.

srowen replied:
Yes, but don't you want the code to handle both cases? and test both cases?

On Wed, Jan 28, 2015 at 4:44 PM, Kirill A. Korinskiy <
notifications@github.com> wrote:

> In core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala
> <https://github.com/apache/spark/pull/4239#discussion_r23700631>:
>
> > @@ -32,7 +32,12 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
> >
> >    def render(request: HttpServletRequest): Seq[Node] = {
> >      listener.synchronized {
> > -      val jobId = request.getParameter("id").toInt
> > +      val paramaterId = request.getParameter("id")
> > +      if (paramaterId.equals("")) {
>
> @srowen <https://github.com/srowen>, yes, you are right.
>
> But for case &id= it's return "". And I used this case for testing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/4239/files#r23700631>.
>

catap replied:
Yes, I updated the code to handle both cases.

BTW I test it only by hand because I not found any testcase for HTTP UI.

andrewor14 commented:
ok to test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26247/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
[Test build #26436 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26436/consoleFull) for   PR 4239 at commit [`b6066e3`](https://github.com/apache/spark/commit/b6066e3d72e95b1c0e53ddc84ad8772a4d5d8a09).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26436 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26436/consoleFull) for   PR 4239 at commit [`b6066e3`](https://github.com/apache/spark/commit/b6066e3d72e95b1c0e53ddc84ad8772a4d5d8a09).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26436/
Test FAILed.

SparkQA commented:
[Test build #26623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26623/consoleFull) for   PR 4239 at commit [`ed4154d`](https://github.com/apache/spark/commit/ed4154d1df85bc5a790e1c078d0891f6dd5619a3).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26623/consoleFull) for   PR 4239 at commit [`ed4154d`](https://github.com/apache/spark/commit/ed4154d1df85bc5a790e1c078d0891f6dd5619a3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26623/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/ui/exec/ExecutorThreadDumpPage.scala:
```diff
@@ -20,11 +20,10 @@ package org.apache.spark.ui.exec
-
 import org.apache.spark.ui.{UIUtils, WebUIPage}
 
+import scala.xml.{Node, Text}
```
These imports were correct where they were I believe.

catap replied:
I use Intellij IDEA and this was changed imports.

srowen commented on core/src/main/scala/org/apache/spark/ui/jobs/PoolPage.scala:
```diff
@@ -19,10 +19,10 @@ package org.apache.spark.ui.jobs
 
-import org.apache.spark.scheduler.{Schedulable, StageInfo}
-import org.apache.spark.ui.{WebUIPage, UIUtils}
+import scala.xml.Node
```
Ditto, I think we put scala after java imports and before other stuff

srowen commented on core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala:
```diff
@@ -32,7 +32,12 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
   def render(request: HttpServletRequest): Seq[Node] = {
     listener.synchronized {
-      val jobId = request.getParameter("id").toInt
+      val parameterId = request.getParameter("id")
```
Elsewhere the code uses `require(condition, message)` to generate an `IllegalArgumentException`. I think that could be tidier here. You can also use `isEmpty`:
`require(parameterId != null && !parameterId.isEmpty, "Missing id parameter")`

catap replied:
Yes, you're right, but `parameterId.nonEmpty` is more clear, isn't it?

srowen replied:
Ah right, you can use that in Scala. Yes.

SparkQA commented:
[Test build #27004 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27004/consoleFull) for   PR 4239 at commit [`3ba74ba`](https://github.com/apache/spark/commit/3ba74bab99617e0f9c55d0f6e104113e0b1314a3).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27004 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27004/consoleFull) for   PR 4239 at commit [`3ba74ba`](https://github.com/apache/spark/commit/3ba74bab99617e0f9c55d0f6e104113e0b1314a3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class MasterStateResponse(`
  * `class LocalSparkCluster(`
  * `   *   (4) the main class for the child`
  * `  case class BoundPortsResponse(actorPort: Int, webUIPort: Int, restPort: Option[Int])`
  * `      throw new SubmitRestMissingFieldException("Main class is missing.")`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27004/
Test PASSed.

srowen commented:
It's looking good to me but can you make a JIRA (https://issues.apache.org/jira/browse/SPARK/) and put it into the title like `SPARK-xxxx [WEB UI] ...` ?

catap committed:
[SPARK-5672][Web UI] Return `HTTP ERROR 400` when have missing args

Spark web UI return `HTTP ERROR 500` when GET arguments is missing. Changed it to return a `HTTP ERROR 400`

SparkQA commented:
[Test build #27020 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27020/consoleFull) for   PR 4239 at commit [`520e180`](https://github.com/apache/spark/commit/520e180580325652e4e9b966326d778dc379b881).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27020 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27020/consoleFull) for   PR 4239 at commit [`520e180`](https://github.com/apache/spark/commit/520e180580325652e4e9b966326d778dc379b881).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27020/
Test PASSed.
